sqlrunner/mssql_log_utils.py

```diff
@@ -0,0 +1,150 @@
+import re
+import time
+from collections import defaultdict
+
+
+def execute_with_log(cursor, query, log):
+    """
+    Execute `query` with `connection` while logging SQL, timing and IO information
+
+
+    """
+    connection = cursor.connection
+    log.info(f"SQL:\n{query}")
+
+    # https://pymssql.readthedocs.io/en/stable/ref/_mssql.html#_mssql.MSSQLConnection.set_msghandler
+    messages = []
+    connection._conn.set_msghandler(lambda *args: messages.append(args[-1]))
+    cursor.execute("SET STATISTICS TIME ON")
+    cursor.execute("SET STATISTICS IO ON")
+    start = time.monotonic()
+
+    # Actually run the query
+    cursor.execute(query)
+
+    duration = time.monotonic() - start
+    timings, table_io = parse_statistics_messages(messages)
+
+    if table_io:
+        log.info(format_table_io(table_io))
+
+    # In order to make the logs visually parseable rather than just a wall of text we
+    # want some visual space between logs for each query. The simplest way to achieve
+    # this is to append some newlines to the last thing we log here.
+    append_str_to_last_value(timings, "\n\n")
```

Comment:
The log output showed `"parse_elapsed_ms": "0\n\n"`, which was odd: I couldn't see how wasn't being cast to an `int`. Then I found this line. I really don't think we need to append anything to anything, because our logs aren't a wall of text.